third_party/move/move-vm/profiler/src/probe.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! A profiler that emits USDT (Userland Statically Defined Tracing) probes.
+//! See [usdt](https://crates.io/crates/usdt).
+
+use crate::{Profiler, ProfilerFunction};
+use move_vm_types::instr::Instruction;
+use std::time::Instant;
+
+#[usdt::provider]
+mod vm_profiler {
+    fn function_entry(function: &str) {}
+    fn function_exit(function: &str, nanos: u64) {}
+
+    fn instruction_entry(instruction: &str) {}
+    fn instruction_exit(instruction: &str, nanos: u64) {}
+}
+
+pub struct ProbeProfiler;
+
+impl Default for ProbeProfiler {
+    fn default() -> Self {
+        usdt::register_probes().expect("Failed to register probes");
+        Self
+    }
+}
+
+impl Profiler for ProbeProfiler {
+    type FnGuard = FunctionProbe;
+    type InstrGuard = InstructionProbe;
+
+    #[inline]
+    fn function<F>(&self, function: &F) -> Self::FnGuard
+    where
+        F: ProfilerFunction,
+    {
+        FunctionProbe::new(function)
+    }
+
+    #[inline]
+    fn instruction(&self, instruction: &Instruction) -> Self::InstrGuard {
+        InstructionProbe::new(instruction)
+    }
+}
+
+pub struct FunctionProbe {
+    function: String,
+    t0: Instant,
+}
+
+impl FunctionProbe {
+    #[must_use]
+    fn new<F>(function: &F) -> Self
+    where
+        F: ProfilerFunction,
+    {
+        let function = function.name();
+
+        vm_profiler::function_entry!(|| &function);
```

Comment:
can we take `(|| &function.name())` in?

Reply:
With the current design, the answer is no, because I need to store the function name in the guard, so I can emit it with the exit probe when the guard is dropped. But I just realized that I don’t actually need the name in the exit probe, so I think I will remove it and fix this.